src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -685,11 +697,21 @@ public ICompilationTracker WithCreateCreationPolicy(bool forceRegeneration)
                 if (state is null)
                     return this;
 
-                // If we're already in the state where we are running generators and skeletons (and we're not forcing
-                // regeneration) we don't need to do anything and can just return ourselves. The next request to create
-                // the compilation will do so fully.
-                if (state.CreationPolicy == desiredCreationPolicy && !forceRegeneration)
-                    return this;
+                // If we're not forcing regeneration, we can bail out from doing work in a few cases.
+                if (!forceRegeneration)
+                {
+                    // First If we're *already* in the state where we are running generators and skeletons we don't need
+                    // to do anything and can just return ourselves. The next request to create the compilation will do
+                    // so fully.
+                    if (state.CreationPolicy == desiredCreationPolicy)
+                        return this;
+
+                    // Second, if we know we are already in a final compilation state where the generated documents were
+                    // produced, then clearly we don't need to do anything.  Nothing changed between then and now, so we
+                    // can reuse the final compilation as is.
+                    if (state is FinalCompilationTrackerState { GeneratedDocumentsUpToDate: true })
+                        return this;
```

Comment:
here's the checking point.  We have been told that there was an external signal to rerun generators (a save or build) and that it was not the user explicitly clicking on the button that says "dump all generator state and rerun".   Since we can see we just did that, we don't need to proceed.